Proofs

tkstanczak opened the pull request:
WIP on proof module

LukaszRozmej commented on src/Nethermind/Nethermind.Core.Test/BlockValidatorTests.cs:
```diff
@@ -35,10 +35,10 @@ public void When_more_uncles_than_allowed_returns_false()
             Assert.True(noiseRemoved);
             
-            bool result = blockValidator.ValidateSuggestedBlock(Build.A.Block.WithOmmers(Build.A.BlockHeader.TestObject).TestObject);
+            bool result = blockValidator.ValidateSuggestedBlock(Builders.Build.A.Block.WithOmmers(Builders.Build.A.BlockHeader.TestObject).TestObject);
```
We now have Prepare.A/Build.A/Builders.Build.A - what are the differences?

tkstanczak replied:
it should be always Build A, i remiced prepare.a

LukaszRozmej commented on src/Nethermind/Nethermind.Evm/Tracing/ITxTracer.cs:
```diff
@@ -16,45 +16,92 @@
+
+        void MarkAsFailed(Address recipient, long gasSpent, byte[] output, string error, Keccak stateRoot = null) => throw new NotSupportedException();
+
+        void StartOperation(int depth, long gas, Instruction opcode, int pc) => throw new NotSupportedException();
```
throwing NotSupportedException from an interface method implementation looks like some kind of code smell

tkstanczak replied:
i will remove it

LukaszRozmej commented on src/Nethermind/Nethermind.Store/trieCases.txt:
```diff
@@ -0,0 +1,48 @@
+For any cases where you need to test various trie layouts these are potential cases that you should test:
+
+Note 1) Nodes that are less than 32 bytes are not referenced by has but included in the referring node.
```
wording around 'has', something is not grammatically correct.

tkstanczak replied:
by hash

LukaszRozmej commented on src/Nethermind/Nethermind.Store/trieCases.txt:
```diff
@@ -0,0 +1,48 @@
+For any cases where you need to test various trie layouts these are potential cases that you should test:
+
+Note 1) Nodes that are less than 32 bytes are not referenced by has but included in the referring node.
+(so if branch would have a child with index 2 which is a leaf and the leaf RLP would be shorter than 32 bytes then the branch would not include an RLP hash as a reference but just include the RLP directly).
```
directly-> in place?

tkstanczak replied:
yes, by val

LukaszRozmej commented on src/Nethermind/Nethermind.JsonRpc/Modules/Proof/ProofModule.cs:
```diff
@@ -0,0 +1,213 @@
+
+            
+            
+            CollectHeaders(proofTxTracer, callResultWithProof);
```
Would change the method to return result and usage to 
callResultWithProof.BlockHeaders = CollectHeaders(proofTxTracer);
would be more readable. Its like that for CollectAccountProofs

tkstanczak replied:
ok, initally both were the other way, will change it

LukaszRozmej commented on src/Nethermind/Nethermind.Blockchain.Test/Synchronization/FastSync/NodeDataDownloaderTests.cs:
```diff
@@ -316,21 +317,25 @@ private static (string Name, Action<StateTree, StateDb, StateDb> Action)[] InitS
-
-        static NodeDataDownloaderTests()
+        
+        [MethodImpl(MethodImplOptions.Synchronized)]
```
Do you want something like https://github.com/nunit/docs/wiki/SetUpFixture-Attribute ?

tkstanczak replied:
hm, yes

tkstanczak replied:
no, actually, setup fixture does not work

LukaszRozmej commented on src/Nethermind/Nethermind.Blockchain.Test/Proofs/TxTrieTests.cs:
```diff
@@ -0,0 +1,356 @@
+        public void Can_collect_proof_with_trie_case_3_modified()
+        {
+            Block block = Build.A.Block.WithTransactions(
+                Build.A.Transaction.TestObject,
```
Enumerable.Range(0, XXX).Select(i => Build.A.Transaction.TestObject).ToArray()

tkstanczak replied:
heh, yes

simon-jentzsch commented:
You are fast!

So I am currently testing it:
## proof_getTransactionByHash 

- works perfekt!

simon-jentzsch commented:
## proof_getTransactionReceipt
- the txProof and the header is correct ( I also verified it )
- the receiptProof is not correct.

If you want, you can compare this with the current result from the incubed-nodes:

```
curl --request POST \
  --url https://in3-v2.slock.it/goerli/nd-1 \
  --header 'content-type: application/json' \
  --data '{
    "jsonrpc": "2.0",
    "method": "eth_getTransactionReceipt",
    "params": [
 "0x2963a6ef0577790565cd8e383dc4d1db5c4cc4a5a48e5bfbfde0363bc6026e3d"
    ],
    "id": 8,
	"in3":{ "verification":"proof"}
  }'
```


Correct ReceiptProof:
```js
"receiptProof": [
        "0xf851a059d4052b76a90784da5169ce635e9aa2d69ed3c0dbd573dcf825cd66e79dd9dc80808080808080a0c33999488d6a44863215849c0741a832432ec58c27e01ad47cf61b118de2507a8080808080808080",
        "0xf85180a0c72b3beebecdd894a5fba8b8a01e24ef4941782d10d31557c87aa8a70ca6212ea0e2c8e733cdff100c96da9deed719f14678db7942d2550b08753eef2d66116a488080808080808080808080808080",
        "0xf901e620b901e2f901df018304593fb9010000000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000040000000008080000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000010001000000000000000000000040000000000000000000000000000000000120000000000000000000000000010000000000000000000000000000f8d5f85894c4e9aee187c02eaddc220364b7aadf767e9983aee1a0390e4e84254dc029c8cc7d308ceb2fc7aa177e9664d22e4a3a14eacb657c7444a010cc3c382b13ad9246b74708d03528d294522c558727bd2ed4a242bfb7cf0c3ff8799411b1e97a96664eb7c75a2395457faf6c0645236de1a083510cfe6e5ecb544ca7ea8cf05792a67b420f52676c81ab22bf5980ba34be04b840000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000001fa2a6"
      ]
```

Currently nethermind gives me :

```
    "receiptProof": [
      "0xebd780c32081c0c32081c0808080808080808080808080808080808080808080c33081c08080808080808080",
      "0xd780c32081c0c32081c08080808080808080808080808080",
      "0xc32081c0"
    ],
```

simon-jentzsch commented:
## proof_call

Here I didn't get a result.

I tested with a very simply call:

```
curl --request POST \
  --url http://192.168.2.99:8545/ \
  --header 'content-type: application/json' \
  --data '{
    "jsonrpc": "2.0",
    "method": "proof_call",
    "params": [
			{"to":"0x5f51e413581dd76759e9eed51e63d14c8d1379c8" ,"data":"0x0ee0599c"},
			"latest"],
    "id": 8
  }'
```

While `eth_call` gives me the correct result: (`0x0000000000000000000000001b57d36cdf855cc256feadf76010bfac55502034`),

but `proof_call`only returns an empty structure:

```js
{
  "id": 8,
  "jsonrpc": "2.0",
  "result": {
    "result": "0x",
    "accounts": [],
    "blockHeaders": []
  }
}
```

tkstanczak commented:
@simon-jentzsch will check it for you, not sure if relevant but what network did you run it on? (goerli / mainnet)

better to test with specific blocks rather than latest so I can run exactly the same calls (latest follows the head block)

tkstanczak commented:
I can see now that you tested with goerli, checking

tkstanczak commented:
@simon-jentzsch found a problem of getTransactionReceipt,
fixing, should be ready soon, let me write a test for it first

tkstanczak commented:
nethermind> proof.getTransactionReceipt("0x2963a6ef0577790565cd8e383dc4d1db5c4cc4a5a48e5bfbfde0363bc6026e3d")
Request complete in 6276.4µs
{
  "receipt": {
    "transactionHash": "0x2963a6ef0577790565cd8e383dc4d1db5c4cc4a5a48e5bfbfde0363bc6026e3d",
    "transactionIndex": "0x1",
    "blockHash": "0x2c7dae6f61e1f3742e197f3ba0770cb7bbdf934a40f33d621936faee60918fc8",
    "blockNumber": "0x1fa2a6",
    "cumulativeGasUsed": "0x4593f",
    "gasUsed": "0x1ebdf",
    "from": "0x367a38dade2a7995b43e99ca2d2cc0ac6c981809",
    "to": "0xc4e9aee187c02eaddc220364b7aadf767e9983ae",
    "logs": [
      {
        "removed": false,
        "logIndex": "0x0",
        "transactionIndex": "0x1",
        "transactionHash": "0x2963a6ef0577790565cd8e383dc4d1db5c4cc4a5a48e5bfbfde0363bc6026e3d",
        "blockHash": "0x2c7dae6f61e1f3742e197f3ba0770cb7bbdf934a40f33d621936faee60918fc8",
        "blockNumber": "0x1fa2a6",
        "address": "0xc4e9aee187c02eaddc220364b7aadf767e9983ae",
        "data": "0x10cc3c382b13ad9246b74708d03528d294522c558727bd2ed4a242bfb7cf0c3f",
        "topics": [
          "0x390e4e84254dc029c8cc7d308ceb2fc7aa177e9664d22e4a3a14eacb657c7444"
        ]
      },
      {
        "removed": false,
        "logIndex": "0x1",
        "transactionIndex": "0x1",
        "transactionHash": "0x2963a6ef0577790565cd8e383dc4d1db5c4cc4a5a48e5bfbfde0363bc6026e3d",
        "blockHash": "0x2c7dae6f61e1f3742e197f3ba0770cb7bbdf934a40f33d621936faee60918fc8",
        "blockNumber": "0x1fa2a6",
        "address": "0x11b1e97a96664eb7c75a2395457faf6c0645236d",
        "data": "0x000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000001fa2a6",
        "topics": [
          "0x83510cfe6e5ecb544ca7ea8cf05792a67b420f52676c81ab22bf5980ba34be04"
        ]
      }
    ],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000040000000008080000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000010001000000000000000000000040000000000000000000000000000000000120000000000000000000000000010000000000000000000000000000",
    "status": "0x1",
    "error": ""
  },
  "txProof": [
    "0xf851a0bf79ab2d332db40f5d29bfcdd569fab47d84f135ee029afb0d65f670478e454c80808080808080a0d9921a83842ea6cc5c3f7d44f40b6031c49d5b1cf3956ebe557b7aaa71b9be688080808080808080",
    "0xf85180a0459945f27f63af203a111096ea0f161ceb5beeca402dd40eefe11b49fe2e5d91a0c720b11be865b19875aead00e3cb22f602d7b02171e84e41c0db1b4c4dd889158080808080808080808080808080",
    "0xf902f320b902eff902ec82032a84773594008312ce6094c4e9aee187c02eaddc220364b7aadf767e9983ae80b902844f32195d7b2ee3c00da8b9db08c00923c3bc9f0f0fa5581a7af3369d46c890efdabe69bb000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010cc3c382b13ad9246b74708d03528d294522c558727bd2ed4a242bfb7cf0c3f000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002da08f343359ad7e9a4778ab49ea9e7bbe39ca18e1d37698b6ef042686519bec41a8a03093506b0c4bfac6f89cd19ee348a4dc0c84061bf0ebf03a13a5df9c09494d1e"
  ],
  "receiptProof": [
    "0xf851a0cf2aa1ce6616439f64f5526225dba0b282dec606d3bff738368f0e0b4edfa03680808080808080a0ceaa351292efc47353761bb747fe5531288d65fa6db58e06dec0d8db245b90638080808080808080",
    "0xf85180a05d77f006df645040ca203d45e7409fce9d815e8577b7df6df6bd08712c9b455ca076e8952ffaa25e2da33be632857926820e1ca9a80ae4d2bec8c1402ea2b6ca5b8080808080808080808080808080",
    "0xf9011020b9010cf901098083259cc0b9010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000c0"
  ]
}

tkstanczak commented:
hmm it looks different let me investigate slightly more

tkstanczak commented:
@simon-jentzsch your receipt proof seems to be taking a different receipt RLP encoding - not the one that is used for receipt root calculation?

tkstanczak commented:
{
  "receipt": {
    "transactionHash": "0x2963a6ef0577790565cd8e383dc4d1db5c4cc4a5a48e5bfbfde0363bc6026e3d",
    "transactionIndex": "0x1",
    "blockHash": "0x2c7dae6f61e1f3742e197f3ba0770cb7bbdf934a40f33d621936faee60918fc8",
    "blockNumber": "0x1fa2a6",
    "cumulativeGasUsed": "0x4593f",
    "gasUsed": "0x1ebdf",
    "from": "0x367a38dade2a7995b43e99ca2d2cc0ac6c981809",
    "to": "0xc4e9aee187c02eaddc220364b7aadf767e9983ae",
    "logs": [
      {
        "removed": false,
        "logIndex": "0x0",
        "transactionIndex": "0x1",
        "transactionHash": "0x2963a6ef0577790565cd8e383dc4d1db5c4cc4a5a48e5bfbfde0363bc6026e3d",
        "blockHash": "0x2c7dae6f61e1f3742e197f3ba0770cb7bbdf934a40f33d621936faee60918fc8",
        "blockNumber": "0x1fa2a6",
        "address": "0xc4e9aee187c02eaddc220364b7aadf767e9983ae",
        "data": "0x10cc3c382b13ad9246b74708d03528d294522c558727bd2ed4a242bfb7cf0c3f",
        "topics": [
          "0x390e4e84254dc029c8cc7d308ceb2fc7aa177e9664d22e4a3a14eacb657c7444"
        ]
      },
      {
        "removed": false,
        "logIndex": "0x1",
        "transactionIndex": "0x1",
        "transactionHash": "0x2963a6ef0577790565cd8e383dc4d1db5c4cc4a5a48e5bfbfde0363bc6026e3d",
        "blockHash": "0x2c7dae6f61e1f3742e197f3ba0770cb7bbdf934a40f33d621936faee60918fc8",
        "blockNumber": "0x1fa2a6",
        "address": "0x11b1e97a96664eb7c75a2395457faf6c0645236d",
        "data": "0x000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000001fa2a6",
        "topics": [
          "0x83510cfe6e5ecb544ca7ea8cf05792a67b420f52676c81ab22bf5980ba34be04"
        ]
      }
    ],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000040000000008080000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000010001000000000000000000000040000000000000000000000000000000000120000000000000000000000000010000000000000000000000000000",
    "status": "0x1",
    "error": ""
  },
  "txProof": [
    "0xf851a0bf79ab2d332db40f5d29bfcdd569fab47d84f135ee029afb0d65f670478e454c80808080808080a0d9921a83842ea6cc5c3f7d44f40b6031c49d5b1cf3956ebe557b7aaa71b9be688080808080808080",
    "0xf85180a0459945f27f63af203a111096ea0f161ceb5beeca402dd40eefe11b49fe2e5d91a0c720b11be865b19875aead00e3cb22f602d7b02171e84e41c0db1b4c4dd889158080808080808080808080808080",
    "0xf902f320b902eff902ec82032a84773594008312ce6094c4e9aee187c02eaddc220364b7aadf767e9983ae80b902844f32195d7b2ee3c00da8b9db08c00923c3bc9f0f0fa5581a7af3369d46c890efdabe69bb000000000000000000000000000000000000000000000000000000000000024000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010cc3c382b13ad9246b74708d03528d294522c558727bd2ed4a242bfb7cf0c3f000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000002da08f343359ad7e9a4778ab49ea9e7bbe39ca18e1d37698b6ef042686519bec41a8a03093506b0c4bfac6f89cd19ee348a4dc0c84061bf0ebf03a13a5df9c09494d1e"
  ],
  "receiptProof": [
    "0xf851a059d4052b76a90784da5169ce635e9aa2d69ed3c0dbd573dcf825cd66e79dd9dc80808080808080a0c33999488d6a44863215849c0741a832432ec58c27e01ad47cf61b118de2507a8080808080808080",
    "0xf85180a0c72b3beebecdd894a5fba8b8a01e24ef4941782d10d31557c87aa8a70ca6212ea0e2c8e733cdff100c96da9deed719f14678db7942d2550b08753eef2d66116a488080808080808080808080808080",
    "0xf901e620b901e2f901df018304593fb9010000000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000040000000008080000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000010001000000000000000000000040000000000000000000000000000000000120000000000000000000000000010000000000000000000000000000f8d5f85894c4e9aee187c02eaddc220364b7aadf767e9983aee1a0390e4e84254dc029c8cc7d308ceb2fc7aa177e9664d22e4a3a14eacb657c7444a010cc3c382b13ad9246b74708d03528d294522c558727bd2ed4a242bfb7cf0c3ff8799411b1e97a96664eb7c75a2395457faf6c0645236de1a083510cfe6e5ecb544ca7ea8cf05792a67b420f52676c81ab22bf5980ba34be04b840000000000000000000000000000000000000000000000000000000000000000500000000000000000000000000000000000000000000000000000000001fa2a6"
  ]
}

simon-jentzsch commented:
perfect, this receiptProof is correct

tkstanczak commented:
nethermind> proof.call({"to":"0x5f51e413581dd76759e9eed51e63d14c8d1379c8" ,"data":"0x0ee0599c"}, "latest")
Request complete in 52375.9µs
{
  "result": "0x0000000000000000000000001b57d36cdf855cc256feadf76010bfac55502034",
  "accounts": [
    {
      "accountProof": [
        "0xf90211a06833762ca42247d475578ae80ba1fadc2b8b9d915ed4bf7ba83b65e250de4d69a08de47302b865aad2337568d45c452a9a7fca8fb13ef24489525b1b807eae382ea09876aae66c49a070e88a9b08ce28dd4e121efedb222df65f86090166207c2c1ea020de8952a4929a2b86173b4ed9767046748b2d413bb90f0843c057ff08c648a4a0762a3488e1d6c0a3ea7cb23bb38bb1eba00533a6e6ed1caba6add2db030fe06aa0f9cab64711e52d748a47cb605f8cecf4ef73086e89d6906c9f3d225338621e7ea0b0c05cfef99c83e9097f73a63a1c71daa50a481034d8ffc69c1a9031b7644dd0a08c3c93c9d1b7075ea03d7f03bf280983ff1bfbc2d308f248b5830e730e9b2978a0004095178cf9aa6f4dc930fcf91028e0c214f4d34f7512d8161815c622b9f5e6a08d571180cbe7bc728d766213923f9455b3fa2a0db98cb7d0781b6ae3d41c27aca0191d627145b433d023fed3f5e1c20cab0e7bfa6e0bf0e91f8676b30557e377fda0e4c8a6d050f26a2408e59534b23a9b99e3d419342472d88b86692a49b9ebae14a0f671206efde0e63f28259abc2c115b320cbd05e19a35439144ae127dff816d88a03ed472c52252d06d811953a8973053eb3f34324c8e6fb0a6df7029f4b367ac77a06c6342fadb7aa5f9595d99745efe307ec7d7e8c71a3ff892c675bf8a54791a02a0d58e6c6bcefd8b4c5beeb6ba1a1b346d67dcf97cca2825f04094b4e1ab1cc18d80",
        "0xf90211a03da7e4808f21fcfd72da6171664ce31d0aea313ca607d203963c56861dc6c163a02d6525f3c440f2772382850236a1c744ac5b8c42ca7133ea00dc635b6d002bf1a0b781ed92e0dadef25e75906bb30534828506330b2b649023a1fe6258acc226aea00bbe578a343033adefd96010ef21b86c6d10e5a4e5b70a8ceaa537cf96294ce4a0fc52d4a54a5c4deba6bb03a2bce3e419317cb07bc9980ee44b3c41eb11faa22da036028e5dec7b840827a5b287198a6dbf615a8be3038d640ffacc76d20e472476a0ef7de5670643696c7bd579575a2da52de2f994927c35825e41d23a646d8c1d6ea0c3ad09427ba3ee8edc74828131489eda1a35e934a5214b0f237af943c173e52ba041b7db718dbb1322878de94b5290855627e8222dd8507d9fda78eaace0cf9f14a065b72589d376fac83539c9fdde2cd36fe874fd9f462629f9e95d59aaf6db51f9a0b34e0c88f9cb815ea52757cb0e6c8d734d8e7e6c444526f1dd272ad729bcf63ea053dd6cc24e398f6ec830cc03c8622745c5d0d67f830accce946cd68e39688bcfa0adb253e6a6d46c938243800fa1fbc0edc3bc1730f0406aa881ccf713bac0d0c7a0007aba81fd2bc426af4d1a731314ebe5fd22605c1ca8a4340acf6235408531dca02601b1180e0ae74a5c3182c751f4c8cc8cea68e58aaa8a0913a4970e7162b46ba0886c67ee9ed0e66325e82edddf91855920fb017b6d6862d1fc65873e5447fb6080",
        "0xf90211a0e61060adb9baaee17574da508702375395c97e555c5b4646d4c83a37bd53189fa07cb6d6fb52f9dd262ef3d6bdb564e6885af26a58c31131dc569ee97433b03df3a0d02cf82a17def3d725e2d32fa427b4d8475a25ff33fc5514c433a4e3d0bf69fca0b4a5f46293eb8b84ad9390e5872f87682fd41cd5ed10829a5d23d96ff823048ea0ed75d5183b7a279d17e882d648293716b98e1c46c78c4339f8c4b7caecc5b3c0a064af601f8c3111a01fcca69db50945404059927537d0279c4c25f7db542c34e4a0467a69e944143a13aa877ac255b3a841f784cf5482619f4c702d393bddec4f8ea05605e4828e0f4ac40eddb0e1a56503be168136eda84c48e7ad31a2e0f704d436a0debc470bf1970a23ef0fc7f58e033b619b878011d53db3c3ed44008a773eb73aa045d0f7ae500155f32ca166c494eadcd669d0510d928e9cfdc02f680ce034d9cea0465001e5df9e3c5c3e33579d82a059d27d1f12e6ca0fffbbd01a8f21fb47094ea03eaa270ea2ab0329055b8167b9beebf549f3ea28504f69e4662596b6f6312321a0bc1353196a8b59f4398131079f5a8377eab2489f18a9e01adb6df17fc1450d1fa09040fbea64d42c2b4defadb1474de075a75da60363c6eae9d31001c6623a123ca08c6b1de4cd7e22c08dd12054cf9ede39d10124c5582299c32407bfdd24f5c11da02a00915fe9c91c0bfb259d7fd87f773e4e98e8bd05b9b1d8228d9cf22cf8c56e80",
        "0xf901b1a08f01a9e844c4289c60e7ad8bd761ee954b5e32b79ccc79e65a0cc2e4a9adcff4a06a4c3d39ae780be1bbd0d09dee25893c17b07f6258b379361a8cfd0cc114d384a0c43ad354938adba6ed3739f3677bf957a52935338a527750bf6c5b093460255f80a04e7641fa8d1544887606f1700657ee683ac32572a589774ea5ca4e0e3e16cc78a0b176f67625d290784c7e801406b91cee33dc05f5b2fd0c19046bcaa0eebe1b2ba0ebcd7b1398c1d4e62112c2c07c3662645ab6c8d9cfa0ed0e3e90b5ddb916dbeaa0f977aa82f44d75ea4b8050d58e3bbb892c84d8a960f94b0771c12820d188594780a0de8255ef640068ed0f8917bcdffadbfef283516ad7e0775dc5cbbd87eda4d6bfa0b2c6e533da23f852a722abf10cb0450ee1d93981f92d59f1ca1bedd4badc5401a0acb7c7fbbeb6010213293ec22b578b24a7bb43364eba9e59c0c3d0940e9d6cdba08188703c2d92f14acc3c57ebde3024c4678b621146b81850962c0e344f9e26c1a0b02e1a3aaf168c844d30fae3561da674eca50eae003c82359d55ef65b80245aba05778a1897eb8954976e3d4d055974d887f2c46ab336089d96eb78b946eef81e08080",
        "0xf8689f20aa17ce3c4706fbd30e1e60d8db7861cd45e9a4b8a87bbcde4d986bec56afb846f8440180a056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421a0c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"
      ],
      "balance": "0x0",
      "codeHash": "0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470",
      "nonce": "0x1",
      "storageHash": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421",
      "storageProof": []
    },
    {
      "accountProof": [
        "0xf90211a06833762ca42247d475578ae80ba1fadc2b8b9d915ed4bf7ba83b65e250de4d69a08de47302b865aad2337568d45c452a9a7fca8fb13ef24489525b1b807eae382ea09876aae66c49a070e88a9b08ce28dd4e121efedb222df65f86090166207c2c1ea020de8952a4929a2b86173b4ed9767046748b2d413bb90f0843c057ff08c648a4a0762a3488e1d6c0a3ea7cb23bb38bb1eba00533a6e6ed1caba6add2db030fe06aa0f9cab64711e52d748a47cb605f8cecf4ef73086e89d6906c9f3d225338621e7ea0b0c05cfef99c83e9097f73a63a1c71daa50a481034d8ffc69c1a9031b7644dd0a08c3c93c9d1b7075ea03d7f03bf280983ff1bfbc2d308f248b5830e730e9b2978a0004095178cf9aa6f4dc930fcf91028e0c214f4d34f7512d8161815c622b9f5e6a08d571180cbe7bc728d766213923f9455b3fa2a0db98cb7d0781b6ae3d41c27aca0191d627145b433d023fed3f5e1c20cab0e7bfa6e0bf0e91f8676b30557e377fda0e4c8a6d050f26a2408e59534b23a9b99e3d419342472d88b86692a49b9ebae14a0f671206efde0e63f28259abc2c115b320cbd05e19a35439144ae127dff816d88a03ed472c52252d06d811953a8973053eb3f34324c8e6fb0a6df7029f4b367ac77a06c6342fadb7aa5f9595d99745efe307ec7d7e8c71a3ff892c675bf8a54791a02a0d58e6c6bcefd8b4c5beeb6ba1a1b346d67dcf97cca2825f04094b4e1ab1cc18d80",
        "0xf90211a0065952a4f206b43b560d5ce0f3e54b6511406d7da89e3b9f02a83eb967f8d261a0397d7d0fd80d5bb780f7bbfd5bda6ef2c4dff05604811d16303cf0aa32fbbe69a056f8296b3cf502e8937f8cf1e9351c6879ce7ef5664d672e6a2097365fb85a67a0df12624ef5ca8ff0692991a05c8cc30df24311c89d5a853489eb6d01b7ed6d22a0e390000c8a77f495c6e1b34413ad64e11b9a50cc398eafe117984dea9a239e4ea0955a01a5fef296a9843ae31597b2105170e2880859ce9829309933d8a9508adca0f19a9b53bf5633a3d5b229fdda065a839e40b70f1638982e73964c8c09991fafa0b3da89b41ce2d76a5439286c583722ecf1a8e044d6e810b1b39a88af7e8c355da04a48acb5b38d6d600309376fc3e306e9da96ebe5b6b06af5833c638fec6c3799a04151b00a60b81d43ad4559985d078239486b687bf383f35b85ef9567560c5a88a09a39263e640ad122af4a1468dfefda56e0d39f374a56dae8ccf9c930dc9031c5a0d8c926277548fa74f6faaf39b467a38725a6320a913ee1a5f1800bf0cea73300a0dcea7f283a5bad062aac289b079fcbade8459f1c27882909b292e01f5df85e0ca0523783282ff8e2ecad31ce174ac6624c0afad9b07b979ab52c92f82f2300d340a0dd2496230c684eb462da1e318f16759bcf1b0e195758cc140ea06e399531641aa0ed94bf3db3c1d8b9846ad82b5857d796bfd52602f8a8762743526283cb6b5eff80",
        "0xf90211a02d24dfc2db8ca0346ffa734f8facfcfc2ed7fcfb5cb6d718b34856f3c47e8dc9a00e099826a16ee37b27cdd0a994421be43cc3291fdf18dfbb89a56f2f570fb8ada03979346775454ce6263d9cdfe3a58912727082ea43d3c6de78df01a4b316e631a073622ca9dd94d6ecf9534f25828284d9f702a7eb5fd4fef75864c7c23c8e66cea0b946ec65fa66e88508995356830ea313be7f63556b0706482353e4db498a26e8a05345b84372c7dcdcbb9529084adc0fd752b9df9c79c04a42c66fa603e5139ac8a0f54139e7220b7066ca0287885fa6c49bfa92fab6fd27cb5c947ff1badb36fabea06739677036bf01d8bdd3bb68fa1ae11b84adc773e3aed96b0418772789f8db60a051098b71dabed9b221fef7da07d58da5f1185ad12c5453e37c80e454893808b3a0a84603ac7e3c128a018718b126641055bdb4ec684efe93e612a58b117accfdd7a0cf3672804bc34412846abe2ea7bcd7c2ae1b613cf0b8c255553dcb1af678d821a00fb2360296a34f92a78074f72dae9a970023b278df2aa716d1082084ea3c156da0febb41787822ffd198a9038f1de3c070817580e367b8db91f47fafe7f6f12203a03eb8aea6e0a5a878699e27f714a33c738aade5812b30129c61e0d040c3e6ad9ba07f0f0ba96540218965dca135f5be114040e1c87905dcde86048819376bdc5faca084f60dad0ac9b6ff4317a082f10f20597c4c7d17b4574519ead474ea73aa61b680",
        "0xf90191a02691bcd5e32c39f4f36d5b6b03f091948eb9ca3df225ddad93eb31b8497e4c1aa05555b963f50d2ebd7844dab184b647215a211ad29cb71deff579b6f25e47c864a0bc2809c3a3720f2b0fae7dea4f7efedb4327a48c151b89b090ab99dee7bdcf4aa01f28521aec2d7a50a466f88281e6e827d5509bc5a77be74f3f704fd91079442da05ed7f0cf73f299255bfc2bdf20b31a709518fb71bdcaa9fb8fd488ad07af28ae80a03ecd2de48b44e54b5b892520291fd2601ddff191ba446997f3587d51e8f68ddfa097a4d6d7f15b05d96ed7511e39aae51032fc3648c15eb4d789ce8916f79a78a18080a0296c2142079e1c5b73bd703978eaa04e3b9123a240f7bc046a9cb701fd064435a0c4eccfd2cf10f909356230aed382c081b38aa600e183cfdcd8848bf6d8e0e7aba023d8b0cdfc5807cf338a56b1260118129ce2d1080dca624d02c1e7a44ef7255fa068665baa882f2e62da7897273881f510f3768c3d9878efdf097085ef3f0b5ac480a0d4524fcdca73a594f45be8e41613995cdd8bd38512c6932f04f6d6787ca4856380",
        "0xf87180a0b026d67b1d64d125f66321f13756da4dbf3ed022f130a2dbfcc3796dc77e21caa05e856a872504286cb6dbb11c85fab4ca263a72cbee0a785053fbbb4f2d6078738080808080a09922b3a2aac38c7bd2332ec05abd047f2fb7858f2d2ae4f6c49bce18b73eae948080808080808080",
        "0xf8679e3860ace08b3156e1e8b3657a41748919bc942fd9e0f3de90d72f6e5bdfc3b846f8440180a02073e72c0128ade1ffa8041e540f59fa99cbede9b9a6e6ddc15da8b6ac8cd713a029140efcd5358d1dd75badfaa179e3df0dd53f17a883a30152d82903305697a1"
      ],
      "balance": "0x0",
      "codeHash": "0x29140efcd5358d1dd75badfaa179e3df0dd53f17a883a30152d82903305697a1",
      "nonce": "0x1",
      "storageHash": "0x2073e72c0128ade1ffa8041e540f59fa99cbede9b9a6e6ddc15da8b6ac8cd713",
      "storageProof": [
        {
          "key": "0xc2575a0e9e593c00f959f8c92f12db2869c3395a3b0502d05e2516446f71f85b",
          "proof": [
            "0xf90211a038c21026d9dc67ac857dd3be61dbcdec7754965aa68f7fd38c380dccda91a066a025f481809205c635bc748d4e78918786405fc0e362b86f0c28414c2ada68f84ba0e9622f7a0fa9ede3828f647e4fc4823ddfa0740388c4277b4b011304fafc1dd8a036116c8887c0dc0123a4cf2245c15715e70b974258903b7356e4433b42f66680a0dcf55ed8f57d2c1b0f878e707eb0e97a569aa7ba7862c50ae77de4d1cc0b5821a00be8b340536029c2414fa57ba8e51d736ca4b65aad1b7a003b20302ccc93f7afa02d9e3c9a3e9c20257da776cc8f5ed75c1f9c6f9cf476de236151ced43b8cc8aaa0773d359edc8e0feb9d2b07bf83c7872bbb1f3d71242c1109bfbb84e0695a14a5a0728871eb13da0b2b5c66c9ec335c7059d60dff1154ae1b1277f1f01452e73949a047b0ab62dd1c8ad2c97d36d4b4d29de3f2d1d6a5ca77318de665f4a73efcddd4a0d6faf5540a2035b8ae9370a1d63a3872c9fd3cd58bda75f8910515622abdddefa0b500471bfcc9d941b168b2dcbad5f530b89504d913323fe8e64ec0fcf1c73d80a0b8a03827b3caf845deda97cd6494a368980b87a128a4da8749cc2170df171bf2a01476e6210b8991c920cd8546cc59ab8d2f109058072b32b08b9ae0594b67aae5a0fe20ee58bf02b16706678f26e39c38b63c3dee6d84815b2d6141f9ee78e0ae5ba0db5aa68f09a96b7f545a1c8eb5ec067cf0e517e70c7d13dabc5d43dceaf9214a80",
            "0xf891a0a3f6b740432392121a84294316faee2a02409fdfca6964bd37aad2ffcd161e3380a0b5b79e35acb388e192e23363d547e4e975c78d2ed34dc4453c4e46ce2758a9aa808080a0bb49128038ac36bc9e25af1385d2a92003c90c0596f3a06d754e77c2726c14d38080808080a0569d0fa9d5bea172d4e01482e746a31edb51f38689bd1ddba99f422d7f7cb84f80808080",
            "0xf7a020575a0e9e593c00f959f8c92f12db2869c3395a3b0502d05e2516446f71f85b95941b57d36cdf855cc256feadf76010bfac55502034"
          ],
          "value": "0x1b57d36cdf855cc256feadf76010bfac55502034"
        }
      ]
    }
  ],
  "blockHeaders": [
    "0xf9025ca0519176d2508605cebd38f4ce5e48d7e78dbd3cb590bbd38b11099f78635cb0a5a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347940000000000000000000000000000000000000000a0d9e102a7120a60e20fb19911c871bb10ce4e036167e7358051dc53c13e16b234a0b4d68c2050ce9cb7e2f81bbb8179a1f01be953ef0267c9218249c5b68f850beca0ea867b0effa03748e7abec881f5d2fbee889a9d81a9947720994a4f4cc1ea398b901000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000100000000040000000000000000000000800000000000000000000008000000000000000000000000000000000000000000000000000000002831fa5bd837a120083012d35845e2f07cbb861726f6e696e6b61697a656e000000000000000000000000000000000000000000739d5bf0cf74f95bca1ead85dfe8549753644e9e96c48dbc11f118dc80903eeb122ba21b71437a220e15c5bcf6c50a3990b03a8ac1919c41a6e6ae897841acfe00a00000000000000000000000000000000000000000000000000000000000000000880000000000000000"
  ]
}

tkstanczak commented:
there is a question here how to treat null in From field of the proof_call
so far in our eth_call implementation we were assigning a Gas price of 20Gwei, gas limit of 90000 if values were not provided.

now, I changed it for proof_call to run from a system account (0xffff...ffffe) with GasPrice of 0 and GasLimit of block.GasLimit

it means that the system account will be included in the AccountProof

tkstanczak commented:
still one problem remaining with the Goerli / Clique snapshot manager after the proof_call was called, investigating

simon-jentzsch commented:
> there is a question here how to treat null in From field of the proof_call
> so far in our eth_call implementation we were assigning a Gas price of 20Gwei, gas limit of 90000 if values were not provided.
> 
> now, I changed it for proof_call to run from a system account (0xffff...ffffe) with GasPrice of 0 and GasLimit of block.GasLimit
> 
> it means that the system account will be included in the AccountProof

I would handle it the same way you handle `eth_call`. (since it will be used to proof a eth_call) 
I suppose the you need a valid Transaction first. That's why you use the system account. But in this case we should filter out the system account from the accountProofs (as long as `from` is `null`) , because this would only increase the payload without any benefit. 
GasPrice : 0 is ok (it's `eth_call`!)
The GasLimit is actually important, since this could create Attacks, if you use a high gasLimit,(but the in3-server already checks this). Too bad there is no standard for the default, but geth uses a configurable gascap.  ( https://github.com/ethereum/go-ethereum/blob/bc01593afbb2f0bdbf7f21adc790f161b2a8b41d/cmd/utils/flags.go#L493 )

tkstanczak commented:
> > there is a question here how to treat null in From field of the proof_call
> > so far in our eth_call implementation we were assigning a Gas price of 20Gwei, gas limit of 90000 if values were not provided.
> > now, I changed it for proof_call to run from a system account (0xffff...ffffe) with GasPrice of 0 and GasLimit of block.GasLimit
> > it means that the system account will be included in the AccountProof
> 
> I would handle it the same way you handle `eth_call`. (since it will be used to proof a eth_call)
> I suppose the you need a valid Transaction first. That's why you use the system account. But in this case we should filter out the system account from the accountProofs (as long as `from` is `null`) , because this would only increase the payload without any benefit.
> GasPrice : 0 is ok (it's `eth_call`!)
> The GasLimit is actually important, since this could create Attacks, if you use a high gasLimit,(but the in3-server already checks this). Too bad there is no standard for the default, but geth uses a configurable gascap. ( https://github.com/ethereum/go-ethereum/blob/bc01593afbb2f0bdbf7f21adc790f161b2a8b41d/cmd/utils/flags.go#L493 )

added a gas cap config, not removing system account since we always have to have some account used and it may be also accessed later
